File: book-talk-be/_api/src/main/kotlin/kr/co/booktalk/ApiResult.kt
package kr.co.booktalk

data class ApiResult<R>(
    val data: R,
    val error: Error? = null
) {
    data class Error(
        val message: String,
        /** 에러 코드 */
        val code: String? = null,
    )
}

/**
 * Converts object to [ApiResult]
 * All api endpoints must return [ApiResult].
 */
fun <R> R.toResult() = ApiResult(this)

File: book-talk-ts/packages/ai/src/api-result.ts
export interface ApiError {
  message: string;
  code?: string;
}

export interface ApiResult<R> {
  data: R | null;
  error?: ApiError;
}

/**
 * Converts object to ApiResult.
 * All api endpoints must return ApiResult.
 */
export function toResult<R>(data: R): ApiResult<R> {
  return { data };
}

export function toErrorResult<R = null>(message: string, code?: string): ApiResult<R> {
  return {
    data: null,
    error: { message, code },
  };
}
